src/v/cluster/topics_frontend.cc

```diff
@@ -195,10 +198,46 @@ ss::future<std::vector<topic_result>> topics_frontend::update_topic_properties(
       });
 }
 
+ss::future<std::error_code> topics_frontend::do_update_data_policy(
+  topic_properties_update& update, model::timeout_clock::time_point timeout) {
+    // Alter config rewrites all properties, so we need to delete dp if op is
+    // set and vaule is nullopt
+    bool is_convert_from_set_to_remove = false;
+    if (
+      update.properties.data_policy.op == incremental_update_operation::set
+      && !update.properties.data_policy.value.has_value()) {
+        update.properties.data_policy.op = incremental_update_operation::remove;
+        is_convert_from_set_to_remove = true;
+    }
+
+    switch (update.properties.data_policy.op) {
+    case incremental_update_operation::set:
+        co_return co_await _dp_frontend.local().create_data_policy(
+          update.tp_ns, update.properties.data_policy.value.value(), timeout);
+    case incremental_update_operation::remove: {
+        auto res = co_await _dp_frontend.local().clear_data_policy(
+          update.tp_ns, timeout);
+        // If alter_config sets std::nullopt for dp, we need to delete this dp,
+        // and return succees even data policy does not exist
+        co_return res == std::error_code(cluster::errc::data_policy_not_exists)
+            && is_convert_from_set_to_remove
+          ? std::error_code(cluster::errc::success)
+          : res;
+    }
+    case incremental_update_operation::none:
+        co_return std::error_code(cluster::errc::success);
+    }
+}
+
 ss::future<topic_result> topics_frontend::do_update_topic_properties(
   topic_properties_update update, model::timeout_clock::time_point timeout) {
     update_topic_properties_cmd cmd(update.tp_ns, update.properties);
     try {
+        auto update_dp_res = co_await do_update_data_policy(update, timeout);
+        if (update_dp_res != std::error_code(cluster::errc::success)) {
+            co_return topic_result(
+              std::move(update.tp_ns), cluster::errc(update_dp_res.value()));
```

Comment:
But, I thoughts we should use result when function can return object, and can return error
do_update_data_policy does not return smth smart (only error code)

Reply:
I'd follow the existing patterns. I'm only pointing out that `std::error_code(...)` constructed like this is not common, and data policy doesn't seem to be doing anything out of the ordinary, so it might not be using the common error handling pattern.